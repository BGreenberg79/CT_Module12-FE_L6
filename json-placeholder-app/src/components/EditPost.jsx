import React, { useEffect, useState } from 'react'
import { Form, Button, Container  } from 'react-bootstrap'
import { useParams } from 'react-router-dom'
import axios from 'axios'


const EditPost = () => {
    const {id} = useParams()

    const fetchSinglePost = async (id) => {
        try{
        const response = await axios.get(`https://jsonplaceholder.typicode.com/posts?id=${id}`)
        return response.data[0]}
        catch(err){ console.error(err)}
    }
    fetchSinglePost(id)

    const [post, setPost] = useState({
        title: '',
        body: '',
        userId: 0


    })
    useEffect(() => {
        if (data) {
            console.log(data)
            setPost({id:data.id,
                title:data.title,
                body: data.body, 
                userId: data.userId})
        } else{
            setPost({id:0,
                title: '', 
                body: '', 
                userId: 0})
        }
    }, [data])


    const editPostAPI = async (post) => {
        const response = await axios.put(`https://jsonplaceholder.typicode.com/posts/${post.id}`,
            {title: post.title,
            body: post.body,
            userId: post.userId}, 
            {headers: {'Content-type': 'application/json; charset=UTF-8',}}
        )
        console.log(response)
        return response.data
    }

    const handleSubmit = (event) => {
        event.preventDefault();
        editPostAPI(post)
    }

    return (
        <Container>
            <h1>Edit Post</h1>
            <Form onSubmit={handleSubmit}>
                <Form.Group className="mb-3" controlId="title">
                    <Form.Label>Title:</Form.Label>
                    <Form.Control
                        type="text"
                        placeholder="Enter title"
                        onChange={(event) =>
                            setPost({ ...post, title: event.target.value })
                        }
                        value={post.title}
                    />
                </Form.Group>

                <Form.Group className="mb-3" controlId="body">
                    <Form.Label>Body:</Form.Label>
                    <Form.Control
                        type="text"
                        placeholder="Enter body"
                        onChange={(event) =>
                            setPost({ ...post, body: event.target.value })
                        }
                        value={post.body}
                    />
                </Form.Group>

                <Form.Group className="mb-3" controlId="userId">
                    <Form.Label>User Id:</Form.Label>
                    <Form.Control
                        type="number"
                        placeholder="Enter User Id"
                        onChange={(event) =>
                            setPost({ ...post, userId: Number(event.target.value) })
                        }
                        value={post.userId}
                    />
                </Form.Group>


                <Button variant="success" type="submit">
                    Edit Post
                </Button>
            </Form>
        </Container>
    )
}

export default EditPost