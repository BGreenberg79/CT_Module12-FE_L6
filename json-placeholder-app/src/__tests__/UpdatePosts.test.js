// import React from 'react';
// import { render, screen, fireEvent, waitFor, act } from '@testing-library/react';
// import { MemoryRouter, Routes, Route } from 'react-router-dom';
// import EditPost from '../components/EditPost';
// import axios from 'axios';

// jest.mock('axios');

// // Mocking window.alert
// window.alert = jest.fn();

// test('EditPost updates the form and sends a PUT request', async () => {
//     const mockPost = { id: 2, title: 'Updated Title', body: 'Updated Body' };
//     axios.put.mockResolvedValueOnce({ data: mockPost });

//     // render(
//     //     <MemoryRouter initialEntries={['/edit/1']}>
//     //         <Routes>
//     //             <Route path="/edit/:id" element={<EditPost />} />
//     //         </Routes>
//     //     </MemoryRouter>
//     // );

//     render(
//         <MemoryRouter>
//             <EditPost id={1} />
//         </MemoryRouter>
//     );

//     fireEvent.change(screen.getByLabelText(/Title:/i), {
//         target: { value: 'Updated Title' },
//     });
//     fireEvent.change(screen.getByLabelText(/Body:/i), {
//         target: { value: 'Updated Body' },
//     });

//     const button = await screen.findByRole('button', { name: /Edit Post/i });
//     fireEvent.click(button);

//     // fireEvent.click(screen.getByRole('button', { name: /Edit/i }));

//     await waitFor(() =>
//         expect(axios.put).toHaveBeenCalledWith(
//             'https://jsonplaceholder.typicode.com/posts/1',
//             { title: 'Updated Title', body: 'Updated Body' },
//             { headers: { 'Content-type': 'application/json; charset=UTF-8' } }
//         )
//     );
// });
import React from 'react';
import axios from 'axios';
import { render, fireEvent, waitFor, screen } from '@testing-library/react';
import EditPost from '../components/EditPost';
import { MemoryRouter } from 'react-router-dom';
import { act } from 'react';
import '@testing-library/jest-dom';

jest.mock('axios');


test('EditPost updates the form and sends a PUT request', async () => {

    window.alert = jest.fn();

    // Mock the axios.get response
    axios.get.mockResolvedValueOnce({
        data: { id: 1, title: 'Sample Title', body: 'Sample Body', userId: 1 },
    });

    // Mock the axios.put response
    axios.put.mockResolvedValueOnce({
        data: { id: 1, title: 'Updated Title', body: 'Updated Body', userId: 1 },
    });

    await act(async () => {
        render(
            <MemoryRouter>
                <EditPost id={1} />
            </MemoryRouter>
        )
    });

    // Ensure the title and body inputs are pre-filled
    expect(screen.getByLabelText(/Title:/i).value).toBe('Sample Title');
    expect(screen.getByLabelText(/Body:/i).value).toBe('Sample Body');

    // Simulate user input
    fireEvent.change(screen.getByLabelText(/Title:/i), { target: { value: 'Updated Title' } });
    fireEvent.change(screen.getByLabelText(/Body:/i), { target: { value: 'Updated Body' } });

    // Simulate form submission
    const button = await screen.findByRole('button', { name: /Update/i })
    fireEvent.click(button);


    // Assert axios.put was called with the correct arguments
    await waitFor(() =>
        expect(axios.put).toHaveBeenCalledWith(
            'https://jsonplaceholder.typicode.com/posts/1',
            { title: 'Updated Title', body: 'Updated Body', userId: 1 },
            { headers: { 'Content-type': 'application/json; charset=UTF-8' } }
        )
    );
});


test('matches the snapshot', async () => {
    axios.get.mockResolvedValueOnce({
        data: { id: 1, title: 'Sample Title', body: 'Sample Body', userId: 1 },
    });

    const { asFragment } = render(
        <MemoryRouter>
            <EditPost id={1} />
        </MemoryRouter>
    );
    await waitFor(() => {
        expect(screen.getByLabelText(/Title:/i).value).toBe('Sample Title');
        expect(screen.getByLabelText(/Body:/i).value).toBe('Sample Body');
    });

    expect(asFragment()).toMatchSnapshot();
});