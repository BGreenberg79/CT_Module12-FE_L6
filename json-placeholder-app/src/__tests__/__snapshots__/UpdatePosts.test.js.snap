// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h1>
      Edit Post
    </h1>
    <form
      class=""
    >
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="title"
        >
          Title:
        </label>
        <input
          class="form-control"
          id="title"
          placeholder="Enter title"
          type="text"
          value="Sample Title"
        />
      </div>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="body"
        >
          Body:
        </label>
        <input
          class="form-control"
          id="body"
          placeholder="Enter body"
          type="text"
          value="Sample Body"
        />
      </div>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="userId"
        >
          User Id:
        </label>
        <input
          class="form-control"
          id="userId"
          placeholder="Enter User Id"
          type="number"
          value="1"
        />
      </div>
      <button
        class="btn btn-success"
        type="submit"
      >
        Update
      </button>
    </form>
  </div>
</DocumentFragment>
`;
